lib/bolt/transport/local/shell.rb

```diff
@@ -1,26 +1,191 @@
+          escalate = sudoable && run_as && @user != run_as
+          use_sudo = escalate && @target.options['run-as-command'].nil?
+
+          if options[:interpreter]
+            if command.is_a?(Array)
+              command.unshift(options[:interpreter])
+            else
+              command = [options[:interpreter], command]
+            end
+          end
+
+          command_str = command.is_a?(String) ? command : Shellwords.shelljoin(command)
+
+          if escalate
+            if use_sudo
+              sudo_flags = ["sudo", "-k", "-S", "-u", run_as, "-p", Sudoable.sudo_prompt]
+              sudo_flags += ["-E"] if options[:environment]
+              sudo_str = Shellwords.shelljoin(sudo_flags)
+              command_str = "#{sudo_str} #{command_str}"
+            else
+              run_as_str = Shellwords.shelljoin(@target.options['run-as-command'] + [run_as])
+              command_str = "#{run_as_str} #{command_str}"
+            end
+          end
+
+          command_arr = options[:environment].nil? ? [command_str] : [options[:environment], command_str]
 
-          result_output = Bolt::Node::Output.new
-          result_output.stdout << stdout unless stdout.nil?
-          result_output.stderr << stderr unless stderr.nil?
-          result_output.exit_code = rc.exitstatus
+          begin
+            # Prepare the variables!
+            result_output = Bolt::Node::Output.new
+            in_buffer = options[:stdin] || ''
+            # Chunks of this size will be read/written in one iteration
+            chunk_size = 4096
+            index = 0
+            checked_sudo = 0
+            value = 0
+
+            inp, out, err, t = Open3.popen3(*command_arr)
+            read_streams = { out => String.new,
+                             err => String.new }
+            write_stream = in_buffer.empty? ? [] : [inp]
+
+            until Process.waitpid(t.pid, Process::WNOHANG)
+              timeout = 0.1
+              # See if we can read from out or err, or write to in
+              ready_read, ready_write, = select(read_streams.keys, write_stream, nil, timeout)
+
+              # Read from out and err
+              ready_read&.each do |stream|
+                begin
+                  if use_sudo && checked_sudo < 2
+                    buffer = stream.readpartial(chunk_size)
+                    # Split on newlines, including the newline
+                    lines = buffer.split(/(?<=[\n])/)
+                    # handle_sudo will return the line if it is not a sudo prompt or error
+                    lines.map! { |line| handle_sudo(inp, line) }
+                    read_streams[stream] << lines.join("")
+                    checked_sudo += 1
+                  else
+                    read_streams[stream] << stream.readpartial(chunk_size)
+                  end
+                rescue EOFError
+                end
+              end
+
+              if ready_write && index < in_buffer.length
```

Comment:
If there was a read stream ready and the write stream was not ready, `ready_write` will be an empty array, but this condition will be true, causing it to attempt to write when the stream is not ready.